rfd/0159-password-status.md

```diff
@@ -0,0 +1,125 @@
+
+To do this, we will extend the [`UserSpecV2`](https://github.com/gravitational/teleport/blob/6103a2b1eff2a6bece80ab138ebe68b4ba041a91/api/proto/teleport/legacy/types/types.proto#L3185) message by adding a `PasswordState` field:
+
+```proto
+message UserSpecV2 {
+  // ...existing fields...
+
+  PasswordState PasswordState = 11 [(gogoproto.jsontag) = "password_state,omitempty"];
+}
+
+enum PasswordState {
+  // Unable to tell whether the password has been configured.
+  PASSWORD_STATE_UNSPECIFIED = 0;
+  // Password is known to be not configured.
+  PASSWORD_STATE_UNSET = 1;
+  // Password is known to be configured.
+  PASSWORD_STATE_SET = 2;
+}
+```
+
+The flag will be stored under the `/web/users/<username>/params` key, along with the rest of `UserSpecV2` fields.
+
+`PASSWORD_STATE_UNSPECIFIED` is deliberately set to 0, since this is the default value when there's no information about the password state in the database.
+
+The state changes to `PASSWORD_STATE_UNSET` whenever another user (or cluster onboarding flow) creates a password reset token for a given user. These are scenarios when the system becomes confident that the user has no password. As a consequence, users created after this change is made will never have their password in an unspecified state.
+
+The state changes to `PASSWORD_STATE_SET` whenever user sets/resets their password or successfully signs in using a password. The last scenario, in particular, will allow us to gradually fill in information about existing users that have passwords.
+
+> [!WARNING]
+> The PasswordState flag should _not_ be used for authentication itself. Attempting to shortcut password verification may open a timing attack vulnerability.
+
+### Returning Password State to the Account Settings UI
+
+The old account settings page had separate paths for managing passwords and MFA devices, so there is no endpoint that would return information about all authentication methods specifically for this page. There is only an endpoint that returns a list of MFA devices.
+
+Since we already return `authType` in `/v1/webapi/sites/<site>/context`, it makes sense to extend the returned data structure with the `PasswordStatus` flag, thus making it appear in the `UserContext` structure on the frontend.
+
+## Alternatives Considered
+
+### Querying the Audit Log
+
+Theoretically, we could reach some degree of confidence about whether the user has a password configured by running a query of the audit events, looking for password authentication events, and mark users who have been authenticating with passwords as having these configured. This, however, has following downsides:
+
+1. It's going to be costly. We don't have a good interface for querying for such events.
+2. It's still not going to be 100% reliable. Since audit data is owned by the administrators, we can't be sure about data retention policy. Some log entries may be missing.
+
+### Using a Boolean For the State Flag
```

Comment:
Using an enum instead of a bool is often a known good practice. A few supporting links, just FYI:

* https://protobuf.dev/programming-guides/api/#never-use-booleans-for-two-states
* https://cloud.google.com/apis/design/design_patterns#bool_vs_enum_vs_string